src/vs/workbench/contrib/chat/common/promptSyntax/service/promptsServiceImpl.ts

```diff
@@ -152,12 +156,168 @@ export class PromptsService extends Disposable implements IPromptsService {
+		const customAgentsFolder = joinPath(this.environmentService.workspaceStorageHome, 'custom-agents-provider');
+
+		// First, load existing cached files from the folder
+		let allUris: URI[] = [];
+		try {
+			const folderExists = await this.fileService.exists(customAgentsFolder);
+			if (folderExists) {
+				const cachedFiles = await this.fileService.resolve(customAgentsFolder);
+				if (cachedFiles.children) {
+					allUris = cachedFiles.children
+						.filter(child => !child.isDirectory && child.name.endsWith('.agent.md'))
+						.map(child => child.resource);
+				}
+			}
+		} catch (e) {
+			this.logger.error(`[listCustomAgentsFromProvider] Failed to load cached custom agents from disk`, e instanceof Error ? e.message : String(e));
+		}
+
+		// Refresh from providers and write to disk
+		if (this.customAgentsProviders.length > 0) {
+			for (const providerEntry of this.customAgentsProviders) {
+				try {
+					const agents = await providerEntry.provideCustomAgents({}, token);
+					if (!agents || token.isCancellationRequested) {
+						continue;
+					}
+
+					for (const agent of agents) {
+						const agentUri = joinPath(customAgentsFolder, `${agent.name}.agent.md`);
+						const promptContent = this.generatePromptContentFromCustomAgent(agent);
+
+						try {
+							await this.fileService.createFolder(customAgentsFolder);
+							await this.fileService.writeFile(agentUri, VSBuffer.fromString(promptContent), { unlock: false });
+							allUris.push(agentUri);
+						} catch (e) {
+							this.logger.error(`[listCustomAgentsFromProvider] Failed to write custom agent content to file: ${agentUri}`, e instanceof Error ? e.message : String(e));
+						}
+					}
+				} catch (e) {
+					this.logger.error(`[listCustomAgentsFromProvider] Failed to get custom agents from provider`, e instanceof Error ? e.message : String(e));
+				}
+			}
+		}
+
+		// Deduplicate URIs
+		const uniqueUris = Array.from(new Set(allUris.map(uri => uri.toString()))).map(uriStr => URI.parse(uriStr));
+
+		// Map URIs to IPromptPath
+		const providerExtension = this.customAgentsProviders.length > 0 ? this.customAgentsProviders[0].extension : undefined;
+		if (!providerExtension) {
+			return [];
+		}
+
+		const result: IPromptPath[] = [];
+		for (const uri of uniqueUris) {
+			try {
+				const parsed = await this.parseNew(uri, token);
+				result.push({
+					uri,
+					name: parsed.header?.name ?? getCleanPromptName(uri),
+					description: parsed.header?.description ?? '',
+					storage: PromptsStorage.extension,
+					type: PromptsType.agent,
+					extension: providerExtension
+				} satisfies IExtensionPromptPath);
+			} catch (e) {
+				this.logger.error(`[listCustomAgentsFromProvider] Failed to parse custom agent file: ${uri}`, e instanceof Error ? e.message : String(e));
+			}
+		}
+
+		return result;
+	}
+
+	private generatePromptContentFromCustomAgent(agent: any): string {
+		// Generate prompt file content from CustomAgent metadata
```

Comment:
I don't know how we get the content, but I think it needs to be read/generated on the extension side. It could also be argued to cache it on the extension side. Also there we have access to a storage and all we need to send over is the URI, name, description, argument hint, some info where it comes from.